File: C++/Builder.cpp
class Product1{
    public:
    std::vector<std::string> parts_;
    void ListParts()const{
        std::cout << "Product parts: ";
        for (size_t i=0;i<parts_.size();i++){
            if(parts_[i]== parts_.back()){
                std::cout << parts_[i];
            }else{
                std::cout << parts_[i] << ", ";
            }
        }
        std::cout << "\n\n"; 
    }
};

class Builder {
    public:
    virtual ~Builder() {}
    virtual void ProducePartA() const = 0;
    virtual void ProducePartB() const = 0;
    virtual void ProducePartC() const = 0;
}

class ConcreteBuilder1: public Builder {
    private:
    Product1 *product;

    public:
    ConcreteBuilder1() {
        this -> Reset();
    }

    ~ConcreteBuilder1() {
        delete product;
    }

    void Reset() {
        this -> product = new Product1();
    }

    void ProducePartA() const override {
        this -> product -> parts_.push_back("PartA1");
    }

    void ProducePartB() const override {
        this -> product -> parts_.push_back("PartB1");
    }

    void ProducePartC() const override {
        this->product->parts_.push_back("PartC1");
    }

    Product1 *GetProduct() {
        Product1 *result = this -> product;
        this -> Reset();
        return result;
    }
};

class Decorator {
    /*
    @var Builder
    */
   private:
   Builder* builder;

   public:
   void set_builder(Builder *builder) {
    this -> builder = builder;
   }

   void BuildMinimalViableProduct() {
    this -> builder -> ProducePartA();
   }

   void BuildFullFeaturedProduct() {
    this -> builder -> ProducePartA();
    this -> builder -> ProducePartB();
    this -> builder -> ProducePartC();
   }
};

void ClientCode(Decorator &director)
{
    ConcreteBuilder1 *builder = new ConcreteBuilder1();
    director.set_builder(builder);
    std::cout << "Standard basic product: \n";
    director.BuildFullFeaturedProduct();

    Product1 *p = builder -> GetProduct();
    p -> ListParts();
    delete p;

    std::cout << "Standard full featured product:\n";
    director.BuildFullFeaturedProduct();
    p = builder ->GetProduct();
    p -> ListParts();
    delete p;
}

File: C#/Builder.cs
using System;
using System.Collections.Generic;

namespace DesignPattern.Builder.Conceptual
{
    public interface IBuilder
    {
        void BuildPartA();
        void BuildPartB();
        void BuildPartC();
    }

    public class ConcreteBuilder: IBuilder
    {
        private Product _product = new Product();
        public ConcreteBuilder()
        {
            this.Reset();
        }
        public void Reset()
        {
            this._product = new Product();
        }
        public void BuildPartA()
        {
            this._product.Add("PartA");
        }
        public void BuildPartB()
        {
            this._product.Add("PartB");
        }
        public void BuildPartC()
        {
            this._product.Add("PartC");
        }
        public Product GetProduct()
        {
          Product result = this._product;
          this.Reset();
          return result;
        }
    }

    public class Product
    {
        private List<object> _parts = new List<object>();
        public void Add(string part)
        {
            this._parts.Add(part);
        }
        public string ListParts()
        {
            string str = string.Empty;
            for (int i = 0; i < this._parts.Count; i ++)
            {
                str += this._parts[i] + ",";
            }
            str = str.Remove(str.Length - 2);
            return 'Product parts: ' + str + "\n";
        }
    }

    public class Director
    {
        private IBuilder _builder;
        public IBuilder Builder
        {
            set {_builder = value;}
        }

        public void BuildMinimalViableProduct()
        {
            this._builder.BuildPartA();
        }
        
        public void BuildFullFeaturedProduct()
        {
            this._builder.BuildPartA();
            this._builder.BuildPartB();
            this._builder.BuildPartC();
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var director = new Director();
            var builder = new ConcreteBuilder();
            director.Builder = builder;

            director.BuildMinimalViableProduct();
            Console.WriteLine(builder.GetProduct().ListParts());
        }
    }
}